### PR TITLE
Allow ranks to not visit sections for timer output

### PR DESCRIPTION
**Pull request by tnschmid**
**Description of the change:**

If the TimerOutput object is constructed with MPI_COMM_SELF lap times are not synced. If not all ranks visited a timer section an Assert inside print_wall_time_statistics was triggered or it resulted in a deadlock (depending on the MPI implementation). Now, instead of requiring all ranks to visit the same sections a zero Timer is inserted.

**How the resulting error looks like:**
- OpenMPI just deadlocks
- IntelMPI produces errors I don't understand (I'm not sure why 34 is not equal to 34)
```
[ctest] An error occurred in line <945> of file </dss/dsshome1/02/ge96goh2/dealii/source/base/timer.cc> in function
[ctest]     void dealii::TimerOutput::print_wall_time_statistics(MPI_Comm, double) const
[ctest] The violated condition was: 
[ctest]     ::dealii::deal_II_exceptions::internals::compare_for_equality(sections.size(), Utilities::MPI::max(sections.size(), mpi_comm))
[ctest] Additional information: 
[ctest]     Two sizes or dimensions were supposed to be equal, but aren't. They
[ctest]     are 34 and 34.
``` 

**Where I ran into this problem:**

Pre- and post-operations of matrix free loops do not run if a rank does not own any DoFs (I think), which is sometimes the case for me as many cells have FE_Nothing assigned to them.

### CODE REVIEW

**Review by gassmoeller — COMMENTED** [review]
Ah bummer, another MPI problem in the TimerOutput class. I was dealing with a bunch of them a few months back (see eg. #18866 and some others mentioned there).

I generally agree that `print_wall_time_statistics` should work correctly even if some ranks never enter specific sections. But before I dive into the code and try to understand the corner cases, could you:

- add a test that illustrates the problem and ensures we do not break this behavior in the future (you could take `tests/base/timer_09.cc`, make a copy, and modify it so that only some ranks enter the subsection
- could you explain why you need to check `mpi_communicator` at the beginning of the function and add some documentation on the difference between `mpi_comm` and `mpi_communicator`? The documentation of `print_wall_time_statistics` says: `Note that this call only provides useful information when the TimerOutput object is constructed without an MPI_Comm argument, to let individual sections run without being disturbed by barriers.`. Maybe instead of checking, we should ensure that the timer was constructed without an internal mpi communicator if it doesnt provide useful information anyway in that case? Or would that break a lot of models using this class?

File: source/base/timer.cc [hunk 1]
```diff
@@ -1139,6 +1139,50 @@ TimerOutput::reset()
 
 
 
+void
+TimerOutput::sync_sections(const MPI_Comm mpi_comm) const
+{
+  // No sync required
+  if (mpi_comm == MPI_COMM_SELF)
```
Comment (bangerth) [inline source/base/timer.cc]: That's not wrong, but I think what you really want is this:
```suggestion
  if (Utilities::MPI::n_mpi_processes(mpi_comm) == 1)
```
Same below. Point being is that the communicator doesn't *have* to be `MPI_COMM_SELF`, it could be any communicator that has only one rank.

**Review by bangerth — COMMENTED** [review]
I must admit that I'm confused about the case you're considering. Are you saying that you're creating the `TimerOutput` object with a different MPI communicator than you are calling `print_wall_time_statistics()`? Why would you do that?

I will admit that I don't remember why `print_wall_time_statistics()` takes an MPI communicator as argument to begin with. The class has a member variable that is an MPI communicator. We should just be using that instead.

**Comment by kronbichler** [comment]
> Are you saying that you're creating the `TimerOutput` object with a different MPI communicator than you are calling `print_wall_time_statistics()`? Why would you do that?

This is what I usually do, but my mode of use might have been superseded by updates to the classes in the past years. My use case is that I do not want the timers inside `TimerOutput` to do any synchronization (especially global ones such as those implied by `MPI_Allreduce` and similar calls), because I am interested in where the actual communication dependencies hit the compute time. Think about a cases where we do multiple communications, one use case being so relaxation scheme as multigrid smoother, where we communicate with neighbors in the mesh only. In that case, code like https://github.com/dealii/dealii/blob/dc205d0dd91d6844de6890bf1ca0ec37d8c8f919/source/base/timer.cc#L304-L325 that is called transitively via `TimerOutput::enter_subsection()` with https://github.com/dealii/dealii/blob/dc205d0dd91d6844de6890bf1ca0ec37d8c8f919/source/base/timer.cc#L513 that at least in the second time around calls https://github.com/dealii/dealii/blob/dc205d0dd91d6844de6890bf1ca0ec37d8c8f919/source/base/timer.cc#L175-L180 actually appears to do global communication, which I want to avoid. Or to be more precise, I'd like to have timers having negligible influence on run time, so that they can run all the time. The `print_wall_time_statistics()` should then tell me of possible imbalances, even though it is not a full profiler output that theoretically can show me every message. But if I add global reductions in places where I do not already have them, the premise of "negligible influence" is broken.

It could be that the case is already support in other means, in which case I'm more than happy to support the move towards
> I will admit that I don't remember why print_wall_time_statistics() takes an MPI communicator as argument to begin with. The class has a member variable that is an MPI communicator. We should just be using that instead.

As to the topic: Given my understanding, I would agree with immediate fix. I can also try to help to extend the documentation of the `print_wall_time_statistics()` to better reflect the intent I have in mind.

**Comment by tnschmid** [comment]
As already highlighted by @kronbichler constructing the `TimerOutput` without an MPI communicator and afterward calling `print_wall_time_statistics()` is a common case and also what is suggested in the documentation (second variant):
```
 * "The second variant to cope with this issue is print more information about
 * the recorded times to be able to understand this kind of imbalances without
 * actually adding the barriers..."
 * @code
 *   TimerOutput timer (pcout,
 *                      TimerOutput::never,
 *                      TimerOutput::wall_times);
 * @endcode
 * and then call
 * @code
 *   timer.print_wall_time_statistics(MPI_COMM_WORLD);
 * @endcode
``` 
Let me summarize the current behavior:

Variant 1: sync mode with global collectives (The mode this PR is **not** about):
```
 *   TimerOutput timer (MPI_COMM_WORLD,
 *                      pcout,
 *                      TimerOutput::summary,
 *                      TimerOutput::wall_times);

```
In this case we construct the timer output with an MPI communicator and construct timers with `sync_lap_times=true` (if the MPI communicator is not MPI_COMM_SELF) resulting in global barriers. For this case `print_wall_time_statistics()` will not provide useful information as the timers are synced (timings reflect the maximum) as stated in the documentation `Note that this call only provides useful information when the TimerOutput object is constructed without an MPI_Comm argument, to let individual sections run without being disturbed by barriers.`
In this case we can still call `print_wall_time_statistics()`, but this is not the common case.
I added the Assert back in for this case to require the `mpi_comm` to contain the same number of sections on all ranks if running in sync mode. 

Variant 2: non-sync mode without global collectives (The mode this PR is about):
```
 *   TimerOutput timer (pcout,
 *                      TimerOutput::never,
 *                      TimerOutput::wall_times);
 *   timer.print_wall_time_statistics(mpi_comm);
```

In this case we construct timers with MPI_COMM_SELF and `sync_lap_times=false`. For this case `print_wall_time_statistics()` will provide useful information. In this case global communication only happens when we call `print_wall_time_statistics()`. Previously we required all ranks within the `mpi_comm` (Note: not the `mpi_communicator`) to have visited the same sections. This PR changes this behavior to allow ranks to not visit sections (as they spend 0 time inside not visited sections this is perfectly valid) and inserts the appropriate Timers (0 timers) for the not visited sections constructed with MPI_COMM_SELF and `sync_lap_times=false` like all previous timers.

**Comment by tnschmid** [comment]
>could you explain why you need to check mpi_communicator at the beginning of the function and add some documentation on the difference between mpi_comm and mpi_communicator? The documentation of print_wall_time_statistics says: Note that this call only provides useful information when the TimerOutput object is constructed without an MPI_Comm argument, to let individual sections run without being disturbed by barriers.. Maybe instead of checking, we should ensure that the timer was constructed without an internal mpi communicator if it doesnt provide useful information anyway in that case? Or would that break a lot of models using this class?

I don't know if we should only allow to call `print_wall_time_statistics()` on TimerOutput objects that are running in non-sync mode (`mpi_communicator == MPI_COMM_SELF`). I don't know if there is any use case for this. According to the documentation this is not a very meaningful thing to do, but still fully valid.

**Comment by tnschmid** [comment]
@gassmoeller I added the test case.

**Comment by tnschmid** [comment]
The get-cmake step failed: `Error: get-cmake action execution failed: 'Error: Unexpected HTTP response: 504'`

**Comment by bangerth** [comment]
@tnschmid Is this superseded by #19544?

**Comment by tnschmid** [comment]
Superseded by https://github.com/dealii/dealii/pull/19544 -> Closing
